byron/ledger/executable-spec/src/Ledger/Delegation.hs

```diff
@@ -575,3 +589,124 @@ initialEnvFromGenesisKeys ngk chainLength =
     <*> epochGen 0 10
     <*> slotGen 0 100
     <*> CoreGen.k chainLength (chainLength `div` 10)
+
+--------------------------------------------------------------------------------
+-- Shared support functions for Delegation Properties
+--------------------------------------------------------------------------------
+
+delegatorDelegate :: DCert -> (VKeyGenesis, VKey)
+delegatorDelegate = delegator &&& delegate
+
+ratioInt :: Int -> Int -> Double
+ratioInt x y
+  = fromIntegral x / fromIntegral y
+
+-- | Filters the list with the predicate and returns
+-- the ratio of filtered to original list lengths.
+lenRatio :: ([a] -> [b]) -> [a] -> Double
+lenRatio p xs
+  = ratioInt (length (p xs))
+             (length xs)
+
+maxInt :: [Int] -> Int
+maxInt [] = 0
+maxInt xs = List.maximum xs
+
+-- | True if the tuple count (snd item) is >= 2
+multiple :: (a,Int) -> Bool
+multiple = (2 <=) . snd
+
+-- | Count the number of delegates in the given certificates
+delegateCounts :: [DCert] -> [(VKey, Int)]
+delegateCounts certs
+  = fmap delegatorDelegate certs
+    -- Remove duplicated elements, since we're not
+    -- interested in repeated delegations
+    & List.nub
+    -- Select the (unique) delegates
+    & fmap snd
+    -- If we have more than one occurrence of a delegate,
+    -- then we know that there were multiple delegations
+    -- to that delegate
+    & count
+
+-- | Count the number of delegators in the given certificates
+delegatorCounts :: [DCert] -> [(VKeyGenesis, Int)]
```

Comment:
The implementation of `delegatorCounts` is almost identical to that of `delegateCounts`. How about making an abstract function that based on an input argument either `fmap fst`s or `fmap snd`s and then make the two count functions call into that abstract function?

Reply:
good point... done